actions/actions.go

```diff
@@ -0,0 +1,101 @@
+	 settings, schema settings and events in the WIT.
+
+	 The idea here is to provide a simple, yet powerful "publish-subscribe" system that
+	 can connect any "event" in the system to any "action" with a clear decoupling
+	 of events and actions with the goal of making the associations later dynamic and
+	 configurable by the user ("user connects this event to this action"). Think
+	 of a "IFTTT for WIT" (https://en.wikipedia.org/wiki/IFTTT).
+
+	 Actions are generic and atomic execution steps that do exactly one task and
+	 are configurable. The actions system around the actions provide a key-based
+	 execution of the actions.
+
+	 Some examples for an application of this system would be:
+		- closing all children of a parent WI that is being closed (the user connects the
+		"close" attribute change event of a WI to an action that closes all WIs of
+		a matching query).
+		- sending out notifications for mentions on markdown (the system executes an
+		action "send notification" for every mention found in markdown values).
+		- moving all WIs from one iteration to the next in the time sequence when
+		the original iteration is closed.
+
+		For all these automations, the actions system provides a re-usable, flexible
+		and later user configurable way of doing that without creating lots of
+		custom code and/or custom process implementations that are hardcoded in the
+		WIT.
+*/
+
+import (
+	"context"
+
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/fabric8-services/fabric8-wit/actions/rules"
+	"github.com/fabric8-services/fabric8-wit/application"
+	"github.com/fabric8-services/fabric8-wit/convert"
+)
+
+// ExecuteActionsByOldNew executes all actions given in the actionConfigList
+// using the mapped configuration strings and returns the new context entity.
+// It takes the old version and the new version of the context entity, comparing them.
+func ExecuteActionsByOldNew(ctx context.Context, db application.DB, userID uuid.UUID, oldContext convert.ChangeDetector, newContext convert.ChangeDetector, actionConfigList map[string]string) (convert.ChangeDetector, []convert.Change, error) {
+	if oldContext == nil || newContext == nil {
+		return nil, nil, errs.New("execute actions called with nil entities")
+	}
+	contextChanges, err := oldContext.ChangeSet(newContext)
+	if err != nil {
+		return nil, nil, err
+	}
+	return ExecuteActionsByChangeset(ctx, db, userID, newContext, contextChanges, actionConfigList)
+}
+
+// ExecuteActionsByChangeset executes all actions given in the actionConfigs
+// using the mapped configuration strings and returns the new context entity.
+// It takes a []Change that describes the differences between the old and the new context.
+func ExecuteActionsByChangeset(ctx context.Context, db application.DB, userID uuid.UUID, newContext convert.ChangeDetector, contextChanges []convert.Change, actionConfigs map[string]string) (convert.ChangeDetector, []convert.Change, error) {
```

Comment:
Is `userID` really a user ID or an identity ID? In other places we've called it `modifierID` or `creatorID` to give more meaning to it.

Reply:
I'm fine with a generic term here, just wanted to know if this is an ID for the `users` table or for the `identties` table. I bet it is the last one and then we should call it `identityID` instead of `userID`.